### PR TITLE
Error code explanation extra check

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
r? @Mark-Simulacrum

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
CI passed.

File: src/librustc_error_codes/error_codes/E0230.md [hunk 1]
```diff
@@ -3,15 +3,11 @@ message for when a particular trait isn't implemented on a type placed in a
...
+```compile_fail,E0230
 #![feature(rustc_attrs)]
 
-fn foo<T: Index<u8>>(x: T){}
-
-#[rustc_on_unimplemented = "the type `{Self}` cannot be indexed by `{Idx}`"]
-trait Index<Idx> { /* ... */ }
-
-foo(true); // `bool` does not implement `Index<u8>`
+#[rustc_on_unimplemented = "error on `{Self}` with params `<{A},{B}>`"] // error
```
Comment (oli-obk) [inline src/librustc_error_codes/error_codes/E0230.md:9]: This test is a major change to the previous one and doesn't reflect what the text around it says

File: src/librustc_error_codes/error_codes/E0231.md [hunk 2]
```diff
@@ -3,15 +3,11 @@ message for when a particular trait isn't implemented on a type placed in a
...
+```compile_fail,E0231
 #![feature(rustc_attrs)]
 
-fn foo<T: Index<u8>>(x: T){}
-
-#[rustc_on_unimplemented = "the type `{Self}` cannot be indexed by `{Idx}`"]
-trait Index<Idx> { /* ... */ }
-
-foo(true); // `bool` does not implement `Index<u8>`
+#[rustc_on_unimplemented = "error on `{Self}` with params `<{A},{}>`"] // error!
```
Comment (oli-obk) [inline src/librustc_error_codes/error_codes/E0231.md:9]: same here wrt the changes not being reflected in the surrounding text. Why did you change the example?

File: src/librustc_error_codes/error_codes/E0639.md [hunk 3]
```diff
@@ -3,5 +3,17 @@ instantiated from outside of the defining crate as it has been marked
 as `non_exhaustive` and as such more fields/variants may be added in
 future that could cause adverse side effects for this code.
 
+Erroneous code example:
+
+```ignore (not yet an error apparently)
```
Comment (oli-obk) [inline src/librustc_error_codes/error_codes/E0639.md:8]: This only works cross crate, not sure how to do that in a doc test

Comment (oli-obk) [reply src/librustc_error_codes/error_codes/E0639.md:8]: Within the same crate, exhaustiveness is not a problem, that's why it's not erroring

File: src/librustc_error_codes/error_codes/E0714.md [hunk 4]
```diff
@@ -1,5 +1,19 @@
 A `#[marker]` trait contained an associated item.
 
+Erroneous code example:
+
+```ignore (features can't be used outside of nightly)
```
Comment (oli-obk) [inline src/librustc_error_codes/error_codes/E0714.md:5]: are these tests run with stable? can we change that?

File: src/librustc_error_codes/error_codes/E0231.md (hunk 2 above)
Comment (GuillaumeGomez) [reply src/librustc_error_codes/error_codes/E0231.md:9]: Because the current one wasn't failing with the error code `E0231`.

File: src/librustc_error_codes/error_codes/E0639.md (hunk 3 above)
Comment (GuillaumeGomez) [reply src/librustc_error_codes/error_codes/E0639.md:8]: We can just keep it ignored, but I should update the message.

File: src/librustc_error_codes/error_codes/E0714.md (hunk 4 above)
Comment (GuillaumeGomez) [reply src/librustc_error_codes/error_codes/E0714.md:5]: I think they're not, we should test them like the others.

File: src/librustc_error_codes/error_codes/E0230.md (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustc_error_codes/error_codes/E0230.md:9]: It'll need to be updated I guess. But before, it wasn't throwing the correct error code.

File: src/librustc_error_codes/error_codes/E0232.md [hunk 5]
```diff
@@ -3,15 +3,11 @@ message for when a particular trait isn't implemented on a type placed in a
...
-
-#[rustc_on_unimplemented = "the type `{Self}` cannot be indexed by `{Idx}`"]
-trait Index<Idx> { /* ... */ }
-
-foo(true); // `bool` does not implement `Index<u8>`
+#[rustc_on_unimplemented(lorem="")] // error!
+trait BadAnnotation {}
 ```
 
 there will be an error about `bool` not implementing `Index<u8>`, followed by a
```
Comment (oli-obk) [inline src/librustc_error_codes/error_codes/E0232.md:13]: this still has the wrong explanation

File: src/tools/tidy/src/error_codes_check.rs [hunk 6]
```diff
@@ -15,21 +15,58 @@ const WHITELIST: &[&str] = &[
     "E0727", "E0729",
 ];
 
+// Some error codes don't have any tests apparently...
+const IGNORE_EXPLANATION_CHECK: &[&str] = &[
+    "E0570", "E0601", "E0602", "E0639", "E0729", "E0749", "E0750", "E0751",
```
Comment (oli-obk) [inline src/tools/tidy/src/error_codes_check.rs:20]: Open an issue with a list of these error codes and the goal to remove this constant

File: src/librustc_error_codes/error_codes/E0232.md (hunk 5 above)
Comment (GuillaumeGomez) [reply src/librustc_error_codes/error_codes/E0232.md:13]: Yes but it has to be fixed in another PR. ;)

Comment (oli-obk) [reply src/librustc_error_codes/error_codes/E0232.md:13]: ok, can you open an issue about it?

**Comment by oli-obk** [comment]
CI failed, though I'm not sure what's wrong, there's little information. Maybe spurious

@bors r+
